delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/Permissions.scala

```diff
@@ -71,3 +79,118 @@ trait Permissions {
 }
+
+object Permissions {
+
+  private def withPermissions(
+      state: PermissionsState,
+      minimum: Set[Permission]
+  )(permissions: Set[Permission], instant: Instant, subject: Subject): PermissionsState = {
+    state match {
+      case Initial          =>
+        Current(
+          rev = 1L,
+          permissions = permissions ++ minimum,
+          createdAt = instant,
+          createdBy = subject,
+          updatedAt = instant,
+          updatedBy = subject
+        )
+      case current: Current =>
+        current.copy(
+          rev = state.rev + 1,
+          permissions = permissions ++ minimum,
+          updatedAt = instant,
+          updatedBy = subject
+        )
+    }
+  }
+
+  private[delta] def next(
+      minimum: Set[Permission]
+  )(state: PermissionsState, event: PermissionsEvent): PermissionsState = {
+    val apply                                                  = withPermissions(state, minimum) _
+    def appended(e: PermissionsAppended): PermissionsState     =
+      state match {
+        case Initial if e.rev == 1L           => apply(e.permissions, e.instant, e.subject)
+        case s: Current if s.rev + 1 == e.rev => apply(s.permissions ++ e.permissions, e.instant, e.subject)
+        case other                            => other
+      }
+    def replaced(e: PermissionsReplaced): PermissionsState     =
+      state match {
+        case s if s.rev + 1 == e.rev => apply(e.permissions, e.instant, e.subject)
+        case other                   => other
+      }
+    def subtracted(e: PermissionsSubtracted): PermissionsState =
+      state match {
+        case s: Current if s.rev + 1 == e.rev => apply(s.permissions -- e.permissions, e.instant, e.subject)
+        case other                            => other
+      }
+    def deleted(e: PermissionsDeleted): PermissionsState       =
+      state match {
+        case s: Current if s.rev + 1 == e.rev => apply(Set.empty, e.instant, e.subject)
+        case other                            => other
+      }
+    event match {
+      case e: PermissionsAppended   => appended(e)
+      case e: PermissionsReplaced   => replaced(e)
+      case e: PermissionsSubtracted => subtracted(e)
+      case e: PermissionsDeleted    => deleted(e)
+    }
+  }
+
+  private[delta] def evaluate(minimum: Set[Permission])(state: PermissionsState, cmd: PermissionsCommand)(implicit
+      clock: Clock[UIO] = IO.timer.clock
+  ): IO[PermissionsRejection, PermissionsEvent] = {
+    def accept(f: Instant => PermissionsEvent): UIO[PermissionsEvent]            =
+      clock.realTime(TimeUnit.MILLISECONDS).map(rt => f(Instant.ofEpochMilli(rt)))
+    def reject[A <: PermissionsRejection](rejection: A): IO[A, PermissionsEvent] =
+      IO.raiseError(rejection)
+
+    def replace(c: ReplacePermissions): IO[PermissionsRejection, PermissionsEvent]   =
+      if (c.rev != state.rev) reject(IncorrectRev(c.rev, state.rev))
```

Comment:
It may be out of scope for now but we could accumulate rejections.
With https://typelevel.org/cats/datatypes/validated.html it is quite simple to handle
It is not very interesting for permissions but for more complicated cases, it could be very nice.